### ISSUE TITLE
Channel.stdout end will automatically end .stderr / Channel.exit() semantics wrong

### ISSUE THREAD

**Issue by minesworld** [mscdex/ssh2#342]
These are two heavy usage problems for API user.

Ending the .stdout stream while there might be still data to be transferred via .stderr will result to data loss. 

So the API user must implement a logic himself to prevent closing stdout automatically and instead wait that all stderr data might be transferred before calling .end().

This is even worse when using streams as it breaks the normal node conventions on piping. 

---

It might be possible that a process executed on the server closes its stdout and stderr, does some processing and then exits. 

This behaviour couldn't be archived with the current API without breaking node conventions as the docs state that AFTER an close event no further events could be emitted. 

---

Thus I would suggest the following changes: 
- we should make clear that a session on the server side MUST call Channel.exit() .
- on the server shouldn't Writing CHANNEL_CLOSE before .end() have been called on BOTH stdout and stderr streams AND Channel.exit() has been called.

This way the following behaviour will be ensured:
- all data of stdout and stderr will be transferred
- if the server calls Channel.exit() WHILE or even BEFORE data is being transferred the client will be notified immediately as it is now implemented
- if the server calls Channel.exit() AFTER all data has been transferred the exit code won't get lost even if both stdout and stderr have been closed
- a client ONLY needs to react on the 'close' event as this will be the definitive signal that a session is finished, there is an exit code and all data has been transferred

**Comment by gurell** [comment]
I am seeing a closely related issue. I use git to push data over ssh to a server I implement using ssh2 in node. What happens (somewhat simplified) is that.

1. I run "git push ssh://hostname:port/foo.git HEAD"

2. git launches ssh as a subprocess and gets 3 pipes that it can use to communicate with the subprocess.
    2.1 writeable stdin
    2.2 readable stdout
    2.3 readable stderr

3. ssh connects and launches a remote command "git-receive-pack '/foo.git'"

4. git writes data to (2.1)

5. git closes (2.1)
    git has nothing more to write so it closes the pipe
    note that (2.2) and (2.3) are still open so it should still get data back

at this point the ssh2 server receives and end event and it appears it closes (2.3). I can still write to (2.2) but if I try to write to (2.3) it throws a "write after end" error.

In lib/Channel.js line 92 i find: `if (server && self.stderr.writable) self.stderr.end()` which seems to be where stderr gets closed incorrectly

Basically the three channels should be handled independently of each other the same way node's child process has separate stdin, stdout and stderr streams.

**Comment by mscdex** [comment]
Do either of you have a simple, reproducible, *standalone* example that I can use to add as a test case?

**Comment by minesworld** [comment]
checkout my latest feature/bigdata-stream-tests branch and run:

echo 'maxNumber=100000 Exec( O:vStream,E:vStream)<->( O:wStream,E:wStream )' | DEBUG=1 exitOnSourceEnd=1 node test/test-streaming.js -

**Comment by gurell** [comment]
example of failing code

```javascript
var server = require("ssh2").Server;```
var spawn = require('child_process').spawn;

var serverOptions = {
    privateKey: // hard code server ssh key
        "-----BEGIN RSA PRIVATE KEY-----\n" +
        "MIIEpgIBAAKCAQEA4b5KBb3pENEu8NuU1+N/O+Jcs3wezdK3GBkab8Nf6ljM+H0P\n" +
        "kBM+T+lmAbLPs0MN1bJtkNKQT4GVtu6y6aNYWLPmCdvggdeWQCy960JdgfiFAafM\n" +
        "tWvZWfwa8IPfWwg+ywUIOlnzSPEGIJCmEXdq9ZVVKFQwl7jdtpiFPPWJJ/XbZeaS\n" +
        "FFgFxDoAVXVxUWWWliySLncAzg8SC+k9O1TMcS0Ccz0OM/Q3iSe5c80nBF2Ztb4m\n" +
        "iD2pP2vfBq3BLPqArzGC5H3BzPCabsCvjZyHq0yMgSeQBbjggpKc56RZDtvsUMyk\n" +
        "Piw53vx4fcNfsXdYDdy3IH6edGfso4xoIDnVUwIDAQABAoIBAQCWq4mAd5b9sIDi\n" +
        "Jc0coy0i7RzudyOgQMoP/YFVo1N7eLWmDOK8SPvIHoIFFDHo9hJf1W2nwcKsLlOc\n" +
        "+4jSxEGNQtWBMw6P2zLftn00buE2hIdR6YuKc2do1gmtCnejrMcsfjlVqP/maWir\n" +
        "h6Av/EKZk4h5GGgufCcPXpFfdIqM7hy/V6pVJZUZsyxFpeo1+sHZSZtGghJgj7vF\n" +
        "93O9sHQTJVlAaEcaavcSAkcUGWEj9XDS1Ll6/UCho7xg6C4uem83ov3ifVTycjLT\n" +
        "Pr6w0X2gbvuOW352026LMPV8RmhhBIY5rL64aePXTuuZGESlET4igmwTJnDdxgZY\n" +
        "aYtvR0UBAoGBAPtNH8slu5q4oKMYs9rYtFo8Wq9VXpwiy6wX9N5U/mOaVqShTAio\n" +
        "mS52O27cBiySuJwzxFet5JsH/iagEfbrUF/VRlxDjUtycJUN2Bpt2WMlQPUcuCS0\n" +
        "gXz2CkVHqIo8S34/HOl0QIyilPgSNQPys5tk+VTHwJZwRBWIAcdO5tzBAoGBAOX2\n" +
        "1F5U2srp1xu1PBHXvjG18/2+1UWHre9oD5wlNDV64uN1HKYoAND6VapEb1vwUATk\n" +
        "WOXI4TXVLLjsfK764X6nmPkQbBsCTtREg8SLJ5GmVfcJqCYFlEuM0QDBQvVwz3o9\n" +
        "ugcZOugF7L77tzLaD7ICdpswSW0QbM77VHtVXTMTAoGBANb5J+yBj9YQgKJk3F6n\n" +
        "KMlUfgMlVXTdszu5lqinLqMIAXx92uxBRiDf1YHFSKbSsIg3JIwY3snccYabQiLn\n" +
        "M+93I+SDiRh0gmiP2kDpnaBZ2xKTidckbuz2EN0pKr1uTOpnlRUEdL+ZyTPZZ27E\n" +
        "LO8O1te2tUI8MZH1sgfMsGNBAoGBANnLlMSmjvVA1PasBfFMKBjCoAvBUhHg943X\n" +
        "O3mCDtlbwkj0YKbJGkTR4pI48FLd3ZtUfRFb2Ff4xrt5VnkOqMQgjRPORuSpoekN\n" +
        "3tIljNs38xDd93/TJukKu+OAJh94sTewAGC6t3yKJ/eVRKOHvPe1t4kxCUmfFb5O\n" +
        "cGvCpFs7AoGBAO5VXFCSKWnQJY/J6PwWMybFLeq0jCLpb3VY81Agl4uRiFqFihEb\n" +
        "TNkxYFQLSfzjJNIHliuoNY6RwWPWb7h34S5JGEcZcVW4BXVzlxb4LcO5OtuNbkDE\n" +
        "/n+fF4zN0J0sLH+kKqckGJYQqIe0n8b9a2t9joG9FpaaloNMDpc7zXa7\n" +
        "-----END RSA PRIVATE KEY-----"
};

new server(serverOptions, function(client) {
    client.on("authentication", function(context) {
        context.accept(); // accept everything
    });
    client.on("session", function(accept, reject) {
        var session = accept();
        session.once("exec", function(accept, reject, info) {
            var remote = accept();
            remote.stdin.on("data", function(data) {
                console.log("server got: '" + data + "' on stdin"); 
                remote.stdout.write("pong on stdout");
                remote.stderr.write("pong on stderr");
            });
            remote.stdin.on("end", function() {
                console.log("server got end on stdin");
                remote.stdout.write("pong on stdout"); // this works == correct
                remote.stderr.write("pong on stderr"); // this fails == incorrect
            });
        });
    });
}).listen(12345, "127.0.0.1", function() {
    var child = spawn("ssh", ["-p12345", "-o StrictHostKeyChecking no", "-o UserKnownHostsFile /dev/null", "-o LogLevel QUIET", "localhost", "foo"]);
    child.stdout.on("data", function(data) {
        console.log("client got '" + data + "' on stdout");
    });
    child.stderr.on("data", function(data) {
        console.log("client got '" + data + "' on stderr");
        child.stdin.end();
    });
    child.stdin.write("ping");
});
```

**Comment by mscdex** [comment]
@gurell That should be fixed in master with d412778df9a596ccf21da0f52675ada53b595de1. @minesworld Does that fix work for your original issue too?

**Comment by mscdex** [comment]
Closing this for now. If anyone is still encountering this, let me know.
